File: src/main/java/com/joecollins/graphics/screens/generic/PartySummaryScreen.java
package com.joecollins.graphics.screens.generic;

import com.joecollins.bindings.Bindable;
import com.joecollins.bindings.BindableList;
import com.joecollins.bindings.Binding;
import com.joecollins.bindings.BindingReceiver;
import com.joecollins.bindings.IndexedBinding;
import com.joecollins.graphics.components.RegionSummaryFrame;
import com.joecollins.graphics.utils.StandardFont;
import com.joecollins.models.general.Party;
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Component;
import java.awt.Container;
import java.awt.Dimension;
import java.awt.LayoutManager;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.border.EmptyBorder;

public class PartySummaryScreen extends JPanel {

  private PartySummaryScreen(
      JLabel headerLabel,
      RegionSummaryFrame mainFrame,
      List<RegionSummaryFrame> otherFrames,
      int numRows) {
    setBackground(Color.WHITE);
    setLayout(new BorderLayout());
    add(headerLabel, BorderLayout.NORTH);

    JPanel center = new JPanel();
    center.setBackground(Color.WHITE);
    center.setLayout(new Layout(numRows));
    add(center, BorderLayout.CENTER);

    center.add(mainFrame, "main");
    otherFrames.forEach(o -> center.add(o, "other"));
  }

  private static class Layout implements LayoutManager {

    private final int numRows;
    private Component main = new JPanel();
    private List<Component> others = new ArrayList<>();

    private Layout(int numRows) {
      this.numRows = numRows;
    }

    @Override
    public void addLayoutComponent(String name, Component comp) {
      if ("main".equals(name)) main = comp;
      else others.add(comp);
    }

    @Override
    public void removeLayoutComponent(Component comp) {
      if (main == comp) main = new JPanel();
      others.remove(comp);
    }

    @Override
    public Dimension preferredLayoutSize(Container parent) {
      return new Dimension(1024, 512);
    }

    @Override
    public Dimension minimumLayoutSize(Container parent) {
      return new Dimension(128, 64);
    }

    @Override
    public void layoutContainer(Container parent) {
      int numOtherCols = (int) Math.ceil(1.0 * others.size() / numRows);
      int numTotalCols = numRows + numOtherCols;
      double widthPerCol = 1.0 * parent.getWidth() / numTotalCols;
      double heightPerRow = 1.0 * parent.getHeight() / numRows;
      main.setLocation(5, 5);
      main.setSize((int) (numRows * widthPerCol - 10), (int) (numRows * heightPerRow - 10));
      for (int i = 0; i < others.size(); i++) {
        Component other = others.get(i);
        int row = i / numOtherCols;
        int col = i % numOtherCols + numRows;
        other.setLocation((int) (col * widthPerCol + 5), (int) (row * heightPerRow + 5));
        other.setSize((int) (widthPerCol - 10), (int) (heightPerRow - 10));
      }
    }
  }

  public static <T> Builder<T> ofDiff(
      T mainRegion,
      Function<T, Binding<String>> titleFunc,
      Function<T, Binding<Map<Party, Integer>>> seatFunc,
      Function<T, Binding<Map<Party, Integer>>> seatDiffFunc,
      Function<T, Binding<Map<Party, Double>>> votePctFunc,
      Function<T, Binding<Map<Party, Double>>> votePctDiffFunc,
      int numRows) {
    return new Builder<>(
        mainRegion, titleFunc, seatFunc, seatDiffFunc, votePctFunc, votePctDiffFunc, numRows);
  }

  public static <T> Builder<T> ofPrev(
      T mainRegion,
      Function<T, Binding<String>> titleFunc,
      Function<T, Binding<Map<Party, Integer>>> seatFunc,
      Function<T, Binding<Map<Party, Integer>>> seatPrevFunc,
      Function<T, Binding<Map<Party, Double>>> votePctFunc,
      Function<T, Binding<Map<Party, Double>>> votePctPrevFunc,
      int numRows) {
    Function<T, Binding<Map<Party, Integer>>> seatDiffFunc =
        t -> {
          var curr = seatFunc.apply(t);
          var prev = seatPrevFunc.apply(t);
          return curr.merge(
              prev,
              (c, p) ->
                  Stream.concat(c.keySet().stream(), p.keySet().stream())
                      .distinct()
                      .collect(
                          Collectors.toMap(
                              Function.identity(),
                              party -> c.getOrDefault(party, 0) - p.getOrDefault(party, 0))));
        };
    Function<T, Binding<Map<Party, Double>>> votePctDiffFunc =
        t -> {
          var curr = votePctFunc.apply(t);
          var prev = votePctPrevFunc.apply(t);
          return curr.merge(
              prev,
              (c, p) ->
                  Stream.concat(c.keySet().stream(), p.keySet().stream())
                      .distinct()
                      .collect(
                          Collectors.toMap(
                              Function.identity(),
                              party -> c.getOrDefault(party, 0.0) - p.getOrDefault(party, 0.0))));
        };
    return new Builder<>(
        mainRegion, titleFunc, seatFunc, seatDiffFunc, votePctFunc, votePctDiffFunc, numRows);
  }

  public static class Builder<T> {
    private T mainRegion;
    private List<T> regions = new ArrayList<>();
    private Function<T, Binding<String>> titleFunc;
    private Function<T, Binding<Map<Party, Integer>>> seatFunc;
    private Function<T, Binding<Map<Party, Integer>>> seatDiffFunc;
    private Function<T, Binding<Map<Party, Double>>> votePctFunc;
    private Function<T, Binding<Map<Party, Double>>> votePctDiffFunc;
    private int numRows;

    private Builder(
        T mainRegion,
        Function<T, Binding<String>> titleFunc,
        Function<T, Binding<Map<Party, Integer>>> seatFunc,
        Function<T, Binding<Map<Party, Integer>>> seatDiffFunc,
        Function<T, Binding<Map<Party, Double>>> votePctFunc,
        Function<T, Binding<Map<Party, Double>>> votePctDiffFunc,
        int numRows) {
      this.mainRegion = mainRegion;
      this.titleFunc = titleFunc;
      this.seatFunc = seatFunc;
      this.seatDiffFunc = seatDiffFunc;
      this.votePctFunc = votePctFunc;
      this.votePctDiffFunc = votePctDiffFunc;
      this.numRows = numRows;
    }

    public Builder<T> withRegion(T region) {
      regions.add(region);
      return this;
    }

    public PartySummaryScreen build(Binding<Party> partyBinding) {
      var party = new BindingReceiver<>(partyBinding);

      JLabel headerLabel = new JLabel();
      headerLabel.setFont(StandardFont.readBoldFont(32));
      headerLabel.setHorizontalAlignment(JLabel.CENTER);
      headerLabel.setBorder(new EmptyBorder(5, 0, -5, 0));
      party.getBinding(p -> p.getName().toUpperCase() + " SUMMARY").bind(headerLabel::setText);
      party.getBinding(Party::getColor).bind(headerLabel::setForeground);

      RegionSummaryFrame mainFrame = createFrame(mainRegion, party);
      List<RegionSummaryFrame> otherFrames =
          regions.stream().map(t -> createFrame(t, party)).collect(Collectors.toList());

      return new PartySummaryScreen(headerLabel, mainFrame, otherFrames, numRows);
    }

    private RegionSummaryFrame createFrame(T region, BindingReceiver<Party> party) {
      SinglePartyInput input = new SinglePartyInput();
      seatFunc.apply(region).bind(input::setSeats);
      seatDiffFunc.apply(region).bind(input::setSeatDiff);
      votePctFunc.apply(region).bind(input::setVotePct);
      votePctDiffFunc.apply(region).bind(input::setVotePctDiff);
      party.getBinding().bind(input::setParty);

      BindableList<List<String>> values = new BindableList<>();
      values.setAll(List.of(List.of("", ""), List.of("", "")));
      Binding.propertyBinding(
              input,
              i -> {
                int seats = i.seats.getOrDefault(i.party, 0);
                int diff = i.seatDiff.getOrDefault(i.party, 0);
                return List.of(
                    String.valueOf(seats),
                    diff == 0 ? "\u00b10" : new DecimalFormat("+0;-0").format(diff));
              },
              SinglePartyInput.Property.SEATS,
              SinglePartyInput.Property.SEAT_DIFF,
              SinglePartyInput.Property.PARTY)
          .bind(p -> values.set(0, p));
      Binding.propertyBinding(
              input,
              i -> {
                double vote = i.votePct.getOrDefault(i.party, 0.0);
                double diff = i.votePctDiff.getOrDefault(i.party, 0.0);
                return List.of(
                    new DecimalFormat("0.0%").format(vote),
                    diff == 0 ? "\u00b10.0%" : new DecimalFormat("+0.0%;-0.0%").format(diff));
              },
              SinglePartyInput.Property.VOTE_PCT,
              SinglePartyInput.Property.VOTE_PCT_DIFF,
              SinglePartyInput.Property.PARTY)
          .bind(p -> values.set(1, p));

      RegionSummaryFrame frame = new RegionSummaryFrame();
      frame.setHeaderBinding(titleFunc.apply(region));
      frame.setBorderColorBinding(party.getBinding(Party::getColor));
      frame.setSummaryColorBinding(party.getBinding(Party::getColor));
      frame.setNumSectionsBinding(() -> 2);
      frame.setSectionHeaderBinding(IndexedBinding.listBinding("SEATS", "POPULAR VOTE"));
      frame.setSectionValueBinding(IndexedBinding.propertyBinding(values, Function.identity()));
      return frame;
    }
  }

  private static class SinglePartyInput extends Bindable<SinglePartyInput.Property> {
    private enum Property {
      SEATS,
      SEAT_DIFF,
      VOTE_PCT,
      VOTE_PCT_DIFF,
      PARTY
    }

    private Map<Party, Integer> seats;
    private Map<Party, Integer> seatDiff;
    private Map<Party, Double> votePct;
    private Map<Party, Double> votePctDiff;
    private Party party;

    public void setSeats(Map<Party, Integer> seats) {
      this.seats = seats;
      onPropertyRefreshed(Property.SEATS);
    }

    public void setSeatDiff(Map<Party, Integer> seatDiff) {
      this.seatDiff = seatDiff;
      onPropertyRefreshed(Property.SEAT_DIFF);
    }

    public void setVotePct(Map<Party, Double> votePct) {
      this.votePct = votePct;
      onPropertyRefreshed(Property.VOTE_PCT);
    }

    public void setVotePctDiff(Map<Party, Double> votePctDiff) {
      this.votePctDiff = votePctDiff;
      onPropertyRefreshed(Property.VOTE_PCT_DIFF);
    }

    public void setParty(Party party) {
      this.party = party;
      onPropertyRefreshed(Property.PARTY);
    }
  }
}

File: src/main/java/com/joecollins/graphics/screens/generic/PartySummaryScreen.kt
package com.joecollins.graphics.screens.generic

import com.joecollins.bindings.Bindable
import com.joecollins.bindings.Binding
import com.joecollins.bindings.BindingReceiver
import com.joecollins.graphics.components.FontSizeAdjustingLabel
import com.joecollins.graphics.components.RegionSummaryFrame
import com.joecollins.graphics.utils.StandardFont.readBoldFont
import com.joecollins.models.general.Party
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Component
import java.awt.Container
import java.awt.Dimension
import java.awt.LayoutManager
import java.text.DecimalFormat
import java.util.ArrayList
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.border.EmptyBorder
import kotlin.math.ceil

class PartySummaryScreen private constructor(
    headerLabel: JLabel,
    mainFrame: RegionSummaryFrame,
    otherFrames: List<RegionSummaryFrame>,
    numRows: Int
) : JPanel() {

    private class Layout constructor(private val numRows: Int) : LayoutManager {
        private var main: Component = JPanel()
        private val others: MutableList<Component> = ArrayList()

        override fun addLayoutComponent(name: String, comp: Component) {
            if ("main" == name) main = comp else others.add(comp)
        }

        override fun removeLayoutComponent(comp: Component) {
            if (main === comp) main = JPanel()
            others.remove(comp)
        }

        override fun preferredLayoutSize(parent: Container): Dimension {
            return Dimension(1024, 512)
        }

        override fun minimumLayoutSize(parent: Container): Dimension {
            return Dimension(128, 64)
        }

        override fun layoutContainer(parent: Container) {
            val numOtherCols = ceil(1.0 * others.size / numRows).toInt()
            val numTotalCols = numRows + numOtherCols
            val widthPerCol = 1.0 * parent.width / numTotalCols
            val heightPerRow = 1.0 * parent.height / numRows
            main.setLocation(5, 5)
            main.setSize((numRows * widthPerCol - 10).toInt(), (numRows * heightPerRow - 10).toInt())
            for (i in others.indices) {
                val other = others[i]
                val row = i / numOtherCols
                val col = i % numOtherCols + numRows
                other.setLocation((col * widthPerCol + 5).toInt(), (row * heightPerRow + 5).toInt())
                other.setSize((widthPerCol - 10).toInt(), (heightPerRow - 10).toInt())
            }
        }
    }

    class Builder<T>(
        private val mainRegion: T,
        private val titleFunc: (T) -> Binding<String>,
        private val numRows: Int
    ) {
        private var seatFunc: ((T) -> Binding<Map<Party, Int>>)? = null
        private var seatDiffFunc: ((T) -> Binding<Map<Party, Int>>)? = null
        private var seatsHeader = "SEATS"
        private var votePctFunc: ((T) -> Binding<Map<Party, Double>>)? = null
        private var votePctDiffFunc: ((T) -> Binding<Map<Party, Double>>)? = null
        private var voteHeader = "POPULAR VOTE"

        private val regions: MutableList<T> = ArrayList()

        fun withSeatAndDiff(
            seatFunc: ((T) -> Binding<Map<Party, Int>>),
            seatDiffFunc: ((T) -> Binding<Map<Party, Int>>),
            seatsHeader: String = "SEATS"
        ): Builder<T> {
            this.seatFunc = seatFunc
            this.seatDiffFunc = seatDiffFunc
            this.seatsHeader = seatsHeader
            return this
        }

        fun withSeatAndPrev(
            seatFunc: ((T) -> Binding<Map<Party, Int>>),
            seatPrevFunc: ((T) -> Binding<Map<Party, Int>>),
            seatsHeader: String = "SEATS"
        ): Builder<T> {
            val seatDiffFunc = { t: T ->
                val curr = seatFunc(t)
                val prev = seatPrevFunc(t)
                curr.merge(prev) { c, p ->
                    sequenceOf(c.keys.asSequence(), p.keys.asSequence()).flatten()
                        .distinct()
                        .associateWith { party: Party -> (c[party] ?: 0) - (p[party] ?: 0) }
                }
            }
            return withSeatAndDiff(seatFunc, seatDiffFunc, seatsHeader)
        }

        fun withVotePctAndDiff(
            votePctFunc: ((T) -> Binding<Map<Party, Double>>),
            votePctDiffFunc: ((T) -> Binding<Map<Party, Double>>),
            voteHeader: String = "POPULAR VOTE"
        ): Builder<T> {
            this.votePctFunc = votePctFunc
            this.votePctDiffFunc = votePctDiffFunc
            this.voteHeader = voteHeader
            return this
        }

        fun withVotePctAndPrev(
            votePctFunc: ((T) -> Binding<Map<Party, Double>>),
            votePctPrevFunc: ((T) -> Binding<Map<Party, Double>>),
            voteHeader: String = "POPULAR VOTE"
        ): Builder<T> {
            val votePctDiffFunc = { t: T ->
                val curr = votePctFunc(t)
                val prev = votePctPrevFunc(t)
                curr.merge(prev) { c, p ->
                    sequenceOf(c.keys.asSequence(), p.keys.asSequence()).flatten()
                        .distinct()
                        .associateWith { party: Party -> (c[party] ?: 0.0) - (p[party] ?: 0.0) }
                }
            }
            return withVotePctAndDiff(votePctFunc, votePctDiffFunc, voteHeader)
        }

        fun withRegion(region: T): Builder<T> {
            regions.add(region)
            return this
        }

        fun build(partyBinding: Binding<Party>): PartySummaryScreen {
            val party = BindingReceiver(partyBinding)
            val headerLabel = FontSizeAdjustingLabel()
            headerLabel.font = readBoldFont(32)
            headerLabel.horizontalAlignment = JLabel.CENTER
            headerLabel.border = EmptyBorder(5, 0, -5, 0)
            party.getBinding { it.name.uppercase() + " SUMMARY" }.bind { headerLabel.text = it }
            party.getBinding(Party::color).bind { headerLabel.foreground = it }
            val mainFrame = createFrame(mainRegion, party)
            val otherFrames = regions.map { createFrame(it, party) }
            return PartySummaryScreen(headerLabel, mainFrame, otherFrames, numRows)
        }

        private fun createFrame(region: T, party: BindingReceiver<Party>): RegionSummaryFrame {
            val input = SinglePartyInput()
            seatFunc?.invoke(region)?.bind { input.seats = it }
            seatDiffFunc?.invoke(region)?.bind { input.seatDiff = it }
            votePctFunc?.invoke(region)?.bind { input.votePct = it }
            votePctDiffFunc?.invoke(region)?.bind { input.votePctDiff = it }
            party.getBinding().bind { input.party = it }
            val seatBinding = Binding.propertyBinding<SinglePartyInput, List<String>, SinglePartyInput.Property>(
                input,
                { i: SinglePartyInput ->
                    val seats = i.seats[i.party] ?: 0
                    val diff = i.seatDiff[i.party] ?: 0
                    listOf(
                        seats.toString(),
                        if (diff == 0) "\u00b10" else DecimalFormat("+0;-0").format(diff)
                    )
                },
                SinglePartyInput.Property.SEATS,
                SinglePartyInput.Property.SEAT_DIFF,
                SinglePartyInput.Property.PARTY
            )
            val voteBinding = Binding.propertyBinding<SinglePartyInput, List<String>, SinglePartyInput.Property>(
                input,
                { i: SinglePartyInput ->
                    val vote = i.votePct[i.party] ?: 0.0
                    val diff = i.votePctDiff[i.party] ?: 0.0
                    listOf(
                        DecimalFormat("0.0%").format(vote),
                        if (diff == 0.0) "\u00b10.0%" else DecimalFormat("+0.0%;-0.0%").format(diff)
                    )
                },
                SinglePartyInput.Property.VOTE_PCT,
                SinglePartyInput.Property.VOTE_PCT_DIFF,
                SinglePartyInput.Property.PARTY
            )
            val (headers, values) = when {
                seatFunc == null -> listOf(voteHeader) to voteBinding.map { listOf(it) }
                votePctFunc == null -> listOf(seatsHeader) to seatBinding.map { listOf(it) }
                else -> listOf(seatsHeader, voteHeader) to seatBinding.merge(voteBinding) { s, v -> listOf(s, v) }
            }
            return RegionSummaryFrame(
                headerBinding = titleFunc(region),
                summaryColorBinding = party.getBinding { it.color },
                sectionsBinding = values.map { value ->
                    value.zip(headers) { v, h -> RegionSummaryFrame.SectionWithoutColor(h, v) }
                }
            )
        }
    }

    private class SinglePartyInput : Bindable<SinglePartyInput, SinglePartyInput.Property>() {
        enum class Property {
            SEATS, SEAT_DIFF, VOTE_PCT, VOTE_PCT_DIFF, PARTY
        }

        private var _seats: Map<Party, Int> = emptyMap()
        private var _seatDiff: Map<Party, Int> = emptyMap()
        private var _votePct: Map<Party, Double> = emptyMap()
        private var _votePctDiff: Map<Party, Double> = emptyMap()
        private var _party: Party? = null

        var seats: Map<Party, Int>
        get() = _seats
        set(seats) {
            this._seats = seats
            onPropertyRefreshed(Property.SEATS)
        }

        var seatDiff: Map<Party, Int>
        get() = _seatDiff
        set(seatDiff) {
            this._seatDiff = seatDiff
            onPropertyRefreshed(Property.SEAT_DIFF)
        }

        var votePct: Map<Party, Double>
        get() = _votePct
        set(votePct) {
            this._votePct = votePct
            onPropertyRefreshed(Property.VOTE_PCT)
        }

        var votePctDiff: Map<Party, Double>
        get() = _votePctDiff
        set(votePctDiff) {
            this._votePctDiff = votePctDiff
            onPropertyRefreshed(Property.VOTE_PCT_DIFF)
        }

        var party: Party?
        get() = _party
        set(party) {
            this._party = party
            onPropertyRefreshed(Property.PARTY)
        }
    }

    companion object {
        @JvmStatic fun <T> of(
            mainRegion: T,
            titleFunc: (T) -> Binding<String>,
            numRows: Int
        ): Builder<T> {
            return Builder(mainRegion, titleFunc, numRows)
        }

        @JvmStatic fun <T> ofDiff(
            mainRegion: T,
            titleFunc: (T) -> Binding<String>,
            seatFunc: (T) -> Binding<Map<Party, Int>>,
            seatDiffFunc: (T) -> Binding<Map<Party, Int>>,
            votePctFunc: (T) -> Binding<Map<Party, Double>>,
            votePctDiffFunc: (T) -> Binding<Map<Party, Double>>,
            numRows: Int
        ): Builder<T> {
            return Builder(mainRegion, titleFunc, numRows)
                .withSeatAndDiff(seatFunc, seatDiffFunc)
                .withVotePctAndDiff(votePctFunc, votePctDiffFunc)
        }

        @JvmStatic fun <T> ofPrev(
            mainRegion: T,
            titleFunc: (T) -> Binding<String>,
            seatFunc: (T) -> Binding<Map<Party, Int>>,
            seatPrevFunc: (T) -> Binding<Map<Party, Int>>,
            votePctFunc: (T) -> Binding<Map<Party, Double>>,
            votePctPrevFunc: (T) -> Binding<Map<Party, Double>>,
            numRows: Int
        ): Builder<T> {
            return Builder(mainRegion, titleFunc, numRows)
                .withSeatAndPrev(seatFunc, seatPrevFunc)
                .withVotePctAndPrev(votePctFunc, votePctPrevFunc)
        }
    }

    init {
        background = Color.WHITE
        layout = BorderLayout()
        add(headerLabel, BorderLayout.NORTH)
        val center = JPanel()
        center.background = Color.WHITE
        center.layout = Layout(numRows)
        add(center, BorderLayout.CENTER)
        center.add(mainFrame, "main")
        otherFrames.forEach { center.add(it, "other") }
    }
}
